src/google/protobuf/descriptor.proto

```diff
@@ -481,7 +490,11 @@ message FieldOptions {
   // For Google-internal migration only. Do not use.
   optional bool weak = 10 [default=false];
 
-
+  // Provides the ability to override the name of the property
+  // generated for this field. This is applied to all properties
+  // and methods to do with this field, including HasFoo, FooCount,
+  // FooList etc.
+  optional string csharp_property_name = 11;
```

Comment:
And to reply to Jisi's question what happens if we have property called Number in class Number:
It's a compiler error:
/usr/local/google/home/jtattermusch/testproject/testproject/Program.cs(23,23): Error CS0542: `testproject.Number.Number': member names cannot be the same as their enclosing type (CS0542) (testproject)